numba/lowering.py

```diff
@@ -1169,7 +1172,22 @@ def alloca_lltype(self, name, lltype):
         # Is user variable?
         is_uservar = not name.startswith('$')
         # Allocate space for variable
-        aptr = cgutils.alloca_once(self.builder, lltype, name=name, zfill=True)
+        aptr = cgutils.alloca_once(self.builder, lltype,
+                                   name=name, zfill=False)
+        # Zero-fill differently if we are in loops
+        loops = self.cfg.in_loops(self._current_block_offset)
+        # In loops?
+        if loops:
+            innermost = loops[0]
+            # Zero fill at all loop-entries.
+            for ent in innermost.entries:
+                with self.builder.goto_block(self.blkmap[ent]):
+                    self.builder.store(lltype(None), aptr)
+        # Not in loops?
```

Comment:
I think that this fixes the immediate reported issue but other problems remain, it is however probably impractical to fix this without more information from deeper analysis .

This, for example, leaks.
```python
import numpy as np
import numba
rtsys = numba.runtime.rtsys

@numba.njit
def f_memory_leak(N):
    sum_vec = 0

    for n in range(N):
        vec = np.zeros(7)
        for n in range(N):
            z = np.zeros(7)
        sum_vec += vec[0] + z[0]

    return sum_vec

print(rtsys.get_allocation_stats())
f_memory_leak(2)
print(rtsys.get_allocation_stats())
```

Reply:
I need to do a quick check on this to see if it is related.  This makes me doubt the validity of the current fix.